src/modules/aws_opensearch/entity/domain.ts

```diff
@@ -0,0 +1,223 @@
+import { Max, Min } from 'class-validator';
+import {
+  Check,
+  Column,
+  Entity,
+  JoinColumn,
+  JoinTable,
+  ManyToMany,
+  ManyToOne,
+  PrimaryGeneratedColumn,
+} from 'typeorm';
+
+import {
+  EBSOptions,
+  OpenSearchPartitionInstanceType,
+  OpenSearchWarmPartitionInstanceType,
+} from '@aws-sdk/client-opensearch';
+
+import { Policy } from '../../../services/canonical-iam-policy';
+import { cloudId } from '../../../services/cloud-id';
+import { AwsRegions } from '../../aws_account/entity';
+import { Certificate } from '../../aws_acm/entity';
+import { SecurityGroup } from '../../aws_security_group/entity';
+import { Subnet } from '../../aws_vpc/entity';
+
+/**
+ * Table to manage AWS OpenSearch domains. AWS OpenSearch supports both OpenSearch and ElasticSearch.
+ *
+ * @see https://docs.aws.amazon.com/opensearch-service/latest/developerguide/what-is.html
+ */
+@Entity()
+export class Domain {
+  /**
+   * @private
+   * Auto-incremented ID field
+   */
+  @PrimaryGeneratedColumn()
+  id?: number;
+
+  /**
+   * @public
+   * Name to identify the domain
+   */
+  @Column({ nullable: false })
+  @cloudId
+  domainName: string;
+
+  /**
+   * @public
+   * Custom endpoint to use for opensearch application
+   */
+  @Column({ nullable: true })
+  customEndpoint?: string;
+
+  /**
+   * @public
+   * ForeignKey to the certificate that will be used with the custom domain
+   */
+  @ManyToOne(() => Certificate, certificate => certificate.id, {
+    nullable: true,
+    eager: true,
+  })
+  @JoinColumn({ name: 'endpoint_certificate_id', referencedColumnName: 'id' })
+  customEndpointCertificate?: Certificate;
+
+  /**
+   * @public
+   * Version of the opensearch application - like OpenSearch_2.3
+   */
+  @Column()
+  @Check(`
+    "version" ~ '^Elasticsearch_[0-9]{1}\.[0-9]{1,2}$|^OpenSearch_[0-9]{1,2}\.[0-9]{1,2}$'
+  `)
+  version: string;
+
+  /**
+   * @public
+   * Number of availability zones the opensearch should operate in - between 1 and 3
+   */
+  @Min(1)
+  @Max(3)
+  @Column({ type: 'int' })
+  availabilityZoneCount: number;
+
+  /**
+   * @public
+   * Instance type that is used for opensearch instances
+   */
+  @Column({ type: 'varchar', enum: OpenSearchPartitionInstanceType })
+  instanceType: OpenSearchPartitionInstanceType;
+
+  /**
+   * @public
+   * Number of instances that'll run opensearch
+   */
+  @Min(1)
+  @Max(80)
```

Comment:
nit: maybe add the constraint in the documentation comment?
```suggestion
  /**
   * @public
   * Number of instances that'll run opensearch. Min: 1 - Max: 80
   */
  @Min(1)
  @Max(80)
```